### DISCUSSION TITLE
grid and dashboard layouts

### DISCUSSION THREAD

**Discussion by RossCampbellDev** [grafana/grafonnet#232] [Q&A]
At present, `row.withPanels([])` only works as intended when used within a `util.grid.makeGrid([])` - otherwise any panels added to the row will simply be invisible.

If trying to make use of `row.withRepeat()` while not using `makeGrid([])`, you have to define your row inside your `dashboard.withPanels([])` array, then append another array to that containing the panels that you _would_ have put inside `row.withPanels([)` (if it made any sense)

This then makes it impossible to combine some static panels (eg occurs only once within a dashboard) with anything that is repeated by use of `row.withRepeat()` because every panel mentioned repeats, wherever it occurs in the `dashboard.withPanels([])` array.

when using `makeGrid([])` we still repeat ALL mentioned panels, not just those within the grid

using v11.0.0

**Comment by Duologic** [comment]
This is sadly how rows in work today in Grafana, it is a bit archaic and makes designing complex dashboard layouts with Grafonnet quite a challenge.

Any 'static' panels need to come _before_ any rows in the `withPanels([])` array, if not they will be considered part of the 'rowpanel' that came before it. In either case it'll take a bit of trial and error to figure out what the Grafana render engine will do.

The `makeGrid` function uses the util functions described in https://grafana.github.io/grafonnet/API/util.html which might be useful if you want to write your own layout strategy.
